flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/rules/logical/ReplaceIntersectWithSemiJoinRule.java

```diff
@@ -0,0 +1,96 @@
+
+import org.apache.calcite.plan.RelOptRuleCall;
+import org.apache.calcite.plan.RelRule;
+import org.apache.calcite.rel.RelNode;
+import org.apache.calcite.rel.core.Intersect;
+import org.apache.calcite.rel.core.JoinRelType;
+import org.apache.calcite.rel.core.RelFactories;
+import org.apache.calcite.rex.RexNode;
+import org.apache.calcite.tools.RelBuilder;
+import org.apache.calcite.util.Util;
+import org.immutables.value.Value;
+
+import java.util.List;
+
+import static org.apache.flink.table.planner.plan.utils.SetOpRewriteUtil.generateEqualsCondition;
+
+/**
+ * Planner rule that replaces distinct {@link org.apache.calcite.rel.core.Intersect} with a distinct
+ * {@link org.apache.calcite.rel.core.Aggregate} on a SEMI {@link org.apache.calcite.rel.core.Join}.
+ *
+ * <p>Only handle the case of input size 2.
+ */
+@Value.Enclosing
+public class ReplaceIntersectWithSemiJoinRule
+        extends RelRule<ReplaceIntersectWithSemiJoinRule.ReplaceIntersectWithSemiJoinRuleConfig> {
+
+    public static final ReplaceIntersectWithSemiJoinRule INSTANCE =
+            ReplaceIntersectWithSemiJoinRule.ReplaceIntersectWithSemiJoinRuleConfig.DEFAULT
+                    .toRule();
+
+    private ReplaceIntersectWithSemiJoinRule(ReplaceIntersectWithSemiJoinRuleConfig config) {
+        super(config);
+    }
+
+    @Override
+    public boolean matches(RelOptRuleCall call) {
+        Intersect intersect = call.rel(0);
+        return !intersect.all && intersect.getInputs().size() == 2;
+    }
+
+    @Override
+    public void onMatch(RelOptRuleCall call) {
+        Intersect intersect = call.rel(0);
+        RelNode left = intersect.getInput(0);
+        RelNode right = intersect.getInput(1);
+
+        RelBuilder relBuilder = call.builder();
+        List<Integer> keys = Util.range(left.getRowType().getFieldCount());
+        List<RexNode> conditions = generateEqualsCondition(relBuilder, left, right, keys);
+
+        relBuilder.push(left);
+        relBuilder.push(right);
+        relBuilder
+                .join(JoinRelType.SEMI, conditions)
+                .aggregate(
+                        relBuilder.groupKey(keys.stream().mapToInt(Integer::intValue).toArray()));
+        RelNode rel = relBuilder.build();
+        call.transformTo(rel);
+    }
+
+    /** Rule configuration. */
+    @Value.Immutable(singleton = false)
+    public interface ReplaceIntersectWithSemiJoinRuleConfig extends RelRule.Config {
+        ReplaceIntersectWithSemiJoinRule.ReplaceIntersectWithSemiJoinRuleConfig DEFAULT =
+                ImmutableReplaceIntersectWithSemiJoinRule.ReplaceIntersectWithSemiJoinRuleConfig
+                        .builder()
+                        .build()
+                        .withOperandSupplier(b0 -> b0.operand(Intersect.class).anyInputs())
+                        .withRelBuilderFactory(RelFactories.LOGICAL_BUILDER)
+                        .withDescription("ReplaceIntersectWithSemiJoinRule");
```

Comment:
very nit: in fact could be left as is
however it also could be moved to builder and then the will not be created extra objects

Reply:
left as is, others do it same